bug: `mv: cannot overwrite './<DIRNAME>': Directory not empty`

### Description

While the configuration documentation says that `output_dir` in the JSON is a relative path, and having nested path will work fine for the Golang part of firmware-action (and for local builds), it breaks for the automatic caching and automatic artifact upload and download.

The problem is that the artifact packing and unpacking does not take into account possibility of nested directories.

For example configuration `output_dir: output/linux/`. This will cause the packing process to just move `output/linux/` to `.firmware-action/artifacts/linux/` instead of `.firmware-action/artifacts/output/linux/`. And as result unpacking will move `.firmware-action/artifacts/linux/` to `./linux/` instead of `./output/linux/`.

| Process | Source | Correct Destination | Actual destination |
|---------|--------|---------------------|--------------------|
| packing | `output/linux/` | `.firmware-action/artifacts/output/linux/` | `.firmware-action/artifacts/linux/` |
| unpacking | `.firmware-action/artifacts/linux/` | `./output/linux/` | `./linux/` |

### Version of firmware-action

v0.17.1

### Environment

GitHub CI

### CPU architecture

x86_64, i386, arm64

### OS

any

### Reproduction URL

_No response_

### Reproduction steps

```bash
Having a nested directory path in `output_dir`, for example `output/linux/` will can cause issues in GitHub CI. This problem does not happen in local builds.
1) If there is directory called `linux` for example (be it git submodule), then the step to automatically restore the build cache will break.
2) This will also break the automatic artifacts, because they depend on correct paths.
```

### Screenshots

```bash

```

### Logs

```bash
Run actions/cache/restore@v4
Cache hit for: firmware-action-linux-test-build-XXXX
Cache Size: ~9 MB (9224037 B)
/usr/bin/tar -xf /home/runner/work/_temp/XXXX/cache.tzst -P -C /home/runner/work/project-dir-X/project-dir-X --use-compress-program unzstd
Cache restored successfully
Cache restored from key: firmware-action-linux-test-build-XXXX
Run # Create temporary directory for merged config
Run if [ -d ".firmware-action/artifacts" ] && [ -n "$(ls -A .firmware-action/artifacts/ 2>/dev/null)" ]; then
mv: cannot overwrite './linux': Directory not empty
Error: Process completed with exit code 1.
...
```

### JSON configuration

```json
{
  "linux": {
    "linux-test-build": {
      "output_dir": "output/linux/",
    }
  }
}
```

--- a/.github/workflows/example.yml
+++ b/.github/workflows/example.yml
@@ -565,6 +565,34 @@ jobs:
           retention-days: 14
   # ANCHOR_END: example_build_prune
 
+  build-universal-nested-output:
+    needs:
+      - changes
+      - skip-check
+    runs-on: ubuntu-latest
+    if: ${{ ! (github.event_name == 'pull_request_review' && github.actor != 'github-actions[bot]') && needs.skip-check.outputs.changes == 'true' }}
+    # Skip if pull_request_review on PR not made by a bot
+    steps:
+      - name: Cleanup
+        run: |
+          rm -rf ./* || true
+          rm -rf ./.??* || true
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: firmware-action
+        uses: ./
+        # uses: 9elements/firmware-action
+        with:
+          config: 'tests/example_config__universal__with_nested_output.json'
+          target: 'universal-example-nested-output'
+          recursive: 'false'
+          compile: ${{ needs.changes.outputs.compile }}
+          debug: 'true'
+          enable-cache: 'true'
+          auto-download-artifacts: 'true'
+          auto-upload-artifacts: 'true'
+
   # Example of running on non-Linux systems
   test-operating-systems:
     strategy:

--- a/action.yml
+++ b/action.yml
@@ -350,7 +350,37 @@ runs:
       if: ${{ steps.cache.outputs.cache-hit == 'true' || inputs.enable-cache == 'true' }}
       run: |
         if [ -d ".firmware-action/artifacts" ] && [ -n "$(ls -A .firmware-action/artifacts/ 2>/dev/null)" ]; then
-          mv .firmware-action/artifacts/* ./
+          echo "::group::Unpacking cached artifacts"
+
+          # Process each artifact directory
+          for ARTIFACT in .firmware-action/artifacts/*; do
+            # Skip if not a directory
+            if [ ! -d "${ARTIFACT}" ]; then
+              continue
+            fi
+
+            # Get the base name of the artifact directory
+            ARTIFACT_NAME=$(basename "${ARTIFACT}")
+
+            # Check if there's a corresponding text file with the original path
+            PATH_FILE=".firmware-action/artifacts/${ARTIFACT_NAME}.txt"
+            if [ -f "${PATH_FILE}" ]; then
+              # Read the original output_dir path from the file
+              ORIGINAL_PATH=$(cat "${PATH_FILE}")
+              echo "Restoring artifact '${ARTIFACT_NAME}' to original path: ${ORIGINAL_PATH}"
+
+              # Create parent directories if they don't exist
+              mkdir -p "$(dirname "${ORIGINAL_PATH}")"
+
+              # Move the artifact to its original location
+              mv "${ARTIFACT}" "${ORIGINAL_PATH}"
+            else
+              # Fallback to old behavior if no path file exists
+              echo "No path information found for '${ARTIFACT_NAME}', moving to root directory"
+              mv "${ARTIFACT}" ./
+            fi
+          done
+          echo "::endgroup::"
         fi
 
     #=================================

--- a/cmd/firmware-action/recipes/recipes.go
+++ b/cmd/firmware-action/recipes/recipes.go
@@ -293,13 +293,24 @@ func Execute(ctx context.Context, target string, config *Config) error {
 				// If in GitHub, copy output directory into 'StatusDir', so that it can be cached
 				//   and also automatically uploaded as artifact
 				dirToCache := modules[target].GetOutputDir()
+				// .GetOutputDir() returns verbatim what is defined in the configuration file, no
+				//   pre-processing or any alterations
+				//   (must be relative path)
 				pathToCache := filepath.Join(ArtifactDir, filepath.Base(dirToCache))
+
 				slog.Debug(
 					fmt.Sprintf("Copying output files to '%s' for caching and artifact upload", StatusDir),
-					slog.String("source", dirToCache),
+					slog.String("Source", dirToCache),
 					slog.String("Destination", pathToCache),
 				)
 				if err := filesystem.CopyDir(dirToCache, pathToCache); err != nil {
+					return err
+				}
+
+				// Store path to the original directory which was cached
+				//   (to be used in the GitHub wrapper for unpacking)
+				dirToCacheFilename := filepath.Join(ArtifactDir, filepath.Base(dirToCache)+".txt")
+				if err := os.WriteFile(dirToCacheFilename, []byte(dirToCache), 0o644); err != nil {
 					return err
 				}
 			}

--- a/tests/example_config.json
+++ b/tests/example_config.json
@@ -50,6 +50,18 @@
       "input_files": null,
       "container_input_dir": "inputs/",
       "build_commands": ["echo 'hello world'", "touch test.txt"]
+    },
+    "universal-example-nested-output": {
+      "depends": null,
+      "sdk_url": "golang:latest",
+      "repo_path": "./",
+      "container_output_dirs": null,
+      "container_output_files": ["test.txt"],
+      "output_dir": "output/universal/",
+      "input_dirs": null,
+      "input_files": null,
+      "container_input_dir": "inputs/",
+      "build_commands": ["echo 'hello world'", "touch test.txt"]
     }
   },
   "edk2": {

--- a/tests/example_config__universal__with_nested_output.json
+++ b/tests/example_config__universal__with_nested_output.json
@@ -1,0 +1,21 @@
+{
+  "universal": {
+    "universal-example-nested-output": {
+      "depends": null,
+      "sdk_url": "golang:latest",
+      "repo_path": "./",
+      "container_output_dirs": null,
+      "container_output_files": [
+        "test.txt"
+      ],
+      "output_dir": "output/universal/",
+      "input_dirs": null,
+      "input_files": null,
+      "container_input_dir": "inputs/",
+      "build_commands": [
+        "echo 'hello world'",
+        "touch test.txt"
+      ]
+    }
+  }
+}

